# Data Science Project
## Airline Ticket Prices in India - Data Analysis & Predictive Modeling

**Project Prepared by:** Diaa Aldein Alsayed Ibrahim Osman  
**Prepared for:** Epsilon AI Institute  

**Background:**  
This dataset contains price (airfare) data for flights between major cities in India. India stands as one of the largest aviation hubs globally.

**Project Overview:**  
In this project, we're diving into the Airline Ticket Prices in India dataset to understand how ticket prices work. Our main goal is to create a smart computer model that can predict ticket prices based on different factors. By using tools like data analytics and machine learning, we want to find helpful information in the data.

**Who Will Benefit:**
This project aims to assist airlines, travel agencies, and travelers. For airlines and agencies, the model can provide insights for effective pricing strategies. Travelers can benefit by gaining a better understanding of how ticket prices are determined, helping them make informed choices and possibly find more affordable options.

**Dataset Description:**  
The dataset comprises 10,683 instances and 11 features. The data is unclean, with missing values, and not all instances and features may be necessary to achieve the project's goals.

**Features Description:**  

1. **Airline:** Name of the airline used for travel.

2. **Date_of_Journey:** Date on which a person traveled.

3. **Source:** Departure location of the flight.

4. **Destination:** Arrival location of the flight.

5. **Route:** This field contains information about the starting and ending locations of the journey in the standard format used by airlines.

6. **Dep_Time:** Departure time of the flight from the starting location.

7. **Arrival_Time:** Arrival time of the flight at the destination.

8. **Duration:** Duration of the flight in hours/minutes.

9. **Total_Stops:** Number of total stops the flight made before landing at the destination.

10. **Additional_Info:** Any additional information about a flight.

11. **Price:** Price of the flight.


## Step 1: EDA Data Cleaning & Feature Engineering.

In [1]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import warnings            
warnings.filterwarnings("ignore") 

In [2]:
# loading & veiwing the dataset
df = pd.read_excel('Data_Train.xlsx')
df

Airline Date_of_Journey    Source Destination  \
0           IndiGo      24/03/2019  Banglore   New Delhi   
1        Air India       1/05/2019   Kolkata    Banglore   
2      Jet Airways       9/06/2019     Delhi      Cochin   
3           IndiGo      12/05/2019   Kolkata    Banglore   
4           IndiGo      01/03/2019  Banglore   New Delhi   
...            ...             ...       ...         ...   
10678     Air Asia       9/04/2019   Kolkata    Banglore   
10679    Air India      27/04/2019   Kolkata    Banglore   
10680  Jet Airways      27/04/2019  Banglore       Delhi   
10681      Vistara      01/03/2019  Banglore   New Delhi   
10682    Air India       9/05/2019     Delhi      Cochin   

                       Route Dep_Time  Arrival_Time Duration Total_Stops  \
0                  BLR → DEL    22:20  01:10 22 Mar   2h 50m    non-stop   
1      CCU → IXR → BBI → BLR    05:50         13:15   7h 25m     2 stops   
2      DEL → LKO → BOM → COK    09:25  04:25 10 Jun      19h     2 stops   
3            CCU → NAG → BLR    18:05         23:30   5h 25m      1 stop   
4            BLR → NAG → DEL    16:50         21:35   4h 45m      1 stop   
...                      ...      ...           ...      ...         ...   
10678              CCU → BLR    19:55         22:25   2h 30m    non-stop   
10679              CCU → BLR    20:45         23:20   2h 35m    non-stop   
10680              BLR → DEL    08:20         11:20       3h    non-stop   
10681              BLR → DEL    11:30         14:10   2h 40m    non-stop   
10682  DEL → GOI → BOM → COK    10:55         19:15   8h 20m     2 stops   

      Additional_Info  Price  
0             No info   3897  
1             No info   7662  
2             No info  13882  
3             No info   6218  
4             No info  13302  
...               ...    ...  
10678         No info   4107  
10679         No info   4145  
10680         No info   7229  
10681         No info  12648  
10682         No info  11753  

[10683 rows x 11 columns]

In [3]:
# checking the shape of the dataset
df.shape

(10683, 11)

In [4]:
# Geitting Information about the Data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10683 entries, 0 to 10682
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   Airline          10683 non-null  object
 1   Date_of_Journey  10683 non-null  object
 2   Source           10683 non-null  object
 3   Destination      10683 non-null  object
 4   Route            10682 non-null  object
 5   Dep_Time         10683 non-null  object
 6   Arrival_Time     10683 non-null  object
 7   Duration         10683 non-null  object
 8   Total_Stops      10682 non-null  object
 9   Additional_Info  10683 non-null  object
 10  Price            10683 non-null  int64 
dtypes: int64(1), object(10)
memory usage: 918.2+ KB


In [5]:
#checking for missing data
df.isnull().sum()

Airline            0
Date_of_Journey    0
Source             0
Destination        0
Route              1
Dep_Time           0
Arrival_Time       0
Duration           0
Total_Stops        1
Additional_Info    0
Price              0
dtype: int64

In [6]:
# descriptive statistics for categorical columns
df.describe(include="O")

Airline Date_of_Journey Source Destination            Route  \
count         10683           10683  10683       10683            10682   
unique           12              44      5           6              128   
top     Jet Airways      18/05/2019  Delhi      Cochin  DEL → BOM → COK   
freq           3849             504   4537        4537             2376   

       Dep_Time Arrival_Time Duration Total_Stops Additional_Info  
count     10683        10683    10683       10682           10683  
unique      222         1343      368           5              10  
top       18:55        19:00   2h 50m      1 stop         No info  
freq        233          423      550        5625            8345

From all of the above we can found that the following features:
- maintain the features name to be all in lower case.
- There is Null values in: only 1 null value in Route and 1 value in Total_Stops we are going to drop all null values. 
- Date_of_Journey data type need to be change from object to datetime.

In [7]:
# maintain the features name to be all in lower case.
df.columns 

Index(['Airline', 'Date_of_Journey', 'Source', 'Destination', 'Route',
       'Dep_Time', 'Arrival_Time', 'Duration', 'Total_Stops',
       'Additional_Info', 'Price'],
      dtype='object')

In [8]:
df.columns = df.columns.str.lower()
df.columns

Index(['airline', 'date_of_journey', 'source', 'destination', 'route',
       'dep_time', 'arrival_time', 'duration', 'total_stops',
       'additional_info', 'price'],
      dtype='object')

In [9]:
# dropping null values rows.
df.dropna(inplace=True)

In [10]:
df.isnull().sum()

airline            0
date_of_journey    0
source             0
destination        0
route              0
dep_time           0
arrival_time       0
duration           0
total_stops        0
additional_info    0
price              0
dtype: int64

In [11]:
#checking for duplication in data 
df.duplicated().sum()

220

In [12]:
# Drop duplicates and reset the index
df = df.drop_duplicates().reset_index(drop=True)

In [13]:
df.duplicated().sum()

0

In [14]:
# changing date_of_journey from object to datetime type
df['date_of_journey'] =  pd.to_datetime(df['date_of_journey'])

In [15]:
df.head()

airline date_of_journey    source destination                  route  \
0       IndiGo      2019-03-24  Banglore   New Delhi              BLR → DEL   
1    Air India      2019-01-05   Kolkata    Banglore  CCU → IXR → BBI → BLR   
2  Jet Airways      2019-09-06     Delhi      Cochin  DEL → LKO → BOM → COK   
3       IndiGo      2019-12-05   Kolkata    Banglore        CCU → NAG → BLR   
4       IndiGo      2019-01-03  Banglore   New Delhi        BLR → NAG → DEL   

  dep_time  arrival_time duration total_stops additional_info  price  
0    22:20  01:10 22 Mar   2h 50m    non-stop         No info   3897  
1    05:50         13:15   7h 25m     2 stops         No info   7662  
2    09:25  04:25 10 Jun      19h     2 stops         No info  13882  
3    18:05         23:30   5h 25m      1 stop         No info   6218  
4    16:50         21:35   4h 45m      1 stop         No info  13302

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10462 entries, 0 to 10461
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   airline          10462 non-null  object        
 1   date_of_journey  10462 non-null  datetime64[ns]
 2   source           10462 non-null  object        
 3   destination      10462 non-null  object        
 4   route            10462 non-null  object        
 5   dep_time         10462 non-null  object        
 6   arrival_time     10462 non-null  object        
 7   duration         10462 non-null  object        
 8   total_stops      10462 non-null  object        
 9   additional_info  10462 non-null  object        
 10  price            10462 non-null  int64         
dtypes: datetime64[ns](1), int64(1), object(9)
memory usage: 899.2+ KB


## Checking features for strange values & Feature Engineering

### 1. airline col:

In [17]:
# checking for strange values
df.airline.value_counts() # looks good

Jet Airways                          3700
IndiGo                               2043
Air India                            1694
Multiple carriers                    1196
SpiceJet                              815
Vistara                               478
Air Asia                              319
GoAir                                 194
Multiple carriers Premium economy      13
Jet Airways Business                    6
Vistara Premium economy                 3
Trujet                                  1
Name: airline, dtype: int64

### 2. date_of_journey col:

In [18]:
df.date_of_journey.value_counts() # looks good

2019-06-06    490
2019-05-18    486
2019-09-06    485
2019-12-06    483
2019-05-21    482
2019-09-05    466
2019-03-21    412
2019-05-15    402
2019-06-03    397
2019-05-27    369
2019-06-27    339
2019-06-24    330
2019-01-06    330
2019-03-06    326
2019-03-24    314
2019-06-15    314
2019-03-03    309
2019-09-03    299
2019-03-27    290
2019-05-24    286
2019-06-05    281
2019-01-05    274
2019-12-05    259
2019-01-04    256
2019-01-03    198
2019-03-15    162
2019-03-18    156
2019-12-03    141
2019-09-04    125
2019-03-04    110
2019-06-21    109
2019-06-18    105
2019-06-04    100
2019-04-27     94
2019-04-24     92
2019-03-05     90
2019-04-15     89
2019-04-21     82
2019-04-18     67
2019-12-04     63
Name: date_of_journey, dtype: int64

* date_of_journey as date time it is not useful but we can create new features from it like year, month, day, day name, then dropping the 'Date_of_Journey column.
* also we can create new features like holidays with 1 refering to holiday and 0 for not holiday. after checking from internet:
    - The standard working days in India is Monday to Friday then weekends days are Saturday and Sunday
    - In India , National holidays vary according to its local state but there’re ones which applied over the whole country like: 
    * January 26 -> Republic Day Celebrates the 1950 adoption of the Constitution of India
    * August 15 -> Independence Day Celebrates the 1947 Independence from the British rule
    * October 2	 -> Gandhi Jayanti Honors Mahatma Gandhi, father of the nation, who was born on October 2, 1869
    * Reference: https://en.wikipedia.org/wiki/Public_holidays_in_India
    
* From the Month Features we can Create New features like Seasons. The climate of India consists of a wide range of weather conditions across a vast geographic scale and varied topography. but the main seasons are:
    * Winter, occurring from December to February.
    * Summer or pre-monsoon season, lasting from March to May.
    * Monsoon or rainy season, lasting from June to September. 
    * Post-monsoon or autumn season, lasting from October to November.
    * Reference: https://en.wikipedia.org/wiki/Climate_of_India#Seasons

In [19]:
# extract year, month, day and day_name from date
df['year'] = df['date_of_journey'].dt.year
df['month'] = df['date_of_journey'].dt.month
df['day'] = df['date_of_journey'].dt.day
df['day_name'] = df['date_of_journey'].dt.day_name()

# drop date_of_jpurney
df.drop('date_of_journey', axis=1, inplace=True)

In [20]:
# add holiday column
condition = [(df['day_name'].isin(['Saturday','Sunday'])), (df['month'] == 1) & (df['day'] == 26), (df['month'] == 8) & (df['day'] == 15), (df['month'] == 10) & (df['day'] == 2)]
value = [1,1,1,1]
df['holiday'] = np.select(condition,value)

In [21]:
df['month'].unique()

array([ 3,  1,  9, 12,  6,  5,  4], dtype=int64)

In [22]:
# Creating function seasons from month feature.
def seasons(r):
    if r in [12,1,2]:
        return 'Winter'
    elif r in [3,4,5]:
        return 'Summer'
    elif r in [6,7,8,9]:
        return 'Rainy'
    elif r in [10,11]:
        return 'Autumn'
    
df['season'] = df['month'].apply(seasons)

In [23]:
df['season'].unique()

array(['Summer', 'Winter', 'Rainy'], dtype=object)

In [24]:
df.head()

airline    source destination                  route dep_time  \
0       IndiGo  Banglore   New Delhi              BLR → DEL    22:20   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR    05:50   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK    09:25   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR    18:05   
4       IndiGo  Banglore   New Delhi        BLR → NAG → DEL    16:50   

   arrival_time duration total_stops additional_info  price  year  month  day  \
0  01:10 22 Mar   2h 50m    non-stop         No info   3897  2019      3   24   
1         13:15   7h 25m     2 stops         No info   7662  2019      1    5   
2  04:25 10 Jun      19h     2 stops         No info  13882  2019      9    6   
3         23:30   5h 25m      1 stop         No info   6218  2019     12    5   
4         21:35   4h 45m      1 stop         No info  13302  2019      1    3   

   day_name  holiday  season  
0    Sunday        1  Summer  
1  Saturday        1  Winter  
2    Friday        0   Rainy  
3  Thursday        0  Winter  
4  Thursday        0  Winter

### 3. source col:

In [25]:
df.source.unique() # looks good

array(['Banglore', 'Kolkata', 'Delhi', 'Chennai', 'Mumbai'], dtype=object)

### 4. destination col:

In [26]:
df.destination.unique()

array(['New Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Delhi', 'Hyderabad'],
      dtype=object)

In [27]:
df[df["destination"] == 'New Delhi'].head()

airline    source destination            route dep_time  arrival_time  \
0       IndiGo  Banglore   New Delhi        BLR → DEL    22:20  01:10 22 Mar   
4       IndiGo  Banglore   New Delhi  BLR → NAG → DEL    16:50         21:35   
6  Jet Airways  Banglore   New Delhi  BLR → BOM → DEL    18:55  10:25 13 Mar   
7  Jet Airways  Banglore   New Delhi  BLR → BOM → DEL    08:00  05:05 02 Mar   
8  Jet Airways  Banglore   New Delhi  BLR → BOM → DEL    08:55  10:25 13 Mar   

  duration total_stops              additional_info  price  year  month  day  \
0   2h 50m    non-stop                      No info   3897  2019      3   24   
4   4h 45m      1 stop                      No info  13302  2019      1    3   
6  15h 30m      1 stop  In-flight meal not included  11087  2019     12    3   
7   21h 5m      1 stop                      No info  22270  2019      1    3   
8  25h 30m      1 stop  In-flight meal not included  11087  2019     12    3   

   day_name  holiday  season  
0    Sunday        1  Summer  
4  Thursday        0  Winter  
6   Tuesday        0  Winter  
7  Thursday        0  Winter  
8   Tuesday        0  Winter

In [28]:
df[df["destination"] == 'Delhi'].head()

airline    source destination      route dep_time arrival_time duration  \
22    IndiGo  Banglore       Delhi  BLR → DEL    04:00        06:50   2h 50m   
23    IndiGo  Banglore       Delhi  BLR → DEL    18:55        21:50   2h 55m   
28   Vistara  Banglore       Delhi  BLR → DEL    09:45        12:35   2h 50m   
32    IndiGo  Banglore       Delhi  BLR → DEL    04:00        06:50   2h 50m   
43  Air Asia  Banglore       Delhi  BLR → DEL    11:10        13:55   2h 45m   

   total_stops additional_info  price  year  month  day   day_name  holiday  \
22    non-stop         No info   3943  2019      3    4     Monday        0   
23    non-stop         No info   4823  2019      1    5   Saturday        1   
28    non-stop         No info   4668  2019      6   18    Tuesday        0   
32    non-stop         No info   4423  2019      6    4    Tuesday        0   
43    non-stop         No info   3383  2019      6    5  Wednesday        0   

    season  
22  Summer  
23  Winter  
28   Rainy  
32   Rainy  
43   Rainy

* We can noticed that New Delhi and Delhi are both referring to the same destination with the same route abbreviation DEL. I am going to replace New Delhi with Delhi.

In [29]:
# changing destination New Delhi values with Delhi
df['destination'] = df['destination'].str.replace('New Delhi','Delhi')

In [30]:
df.destination.unique()

array(['Delhi', 'Banglore', 'Cochin', 'Kolkata', 'Hyderabad'],
      dtype=object)

### 5. route col:

In [31]:
df.route.unique() # looks good

array(['BLR → DEL', 'CCU → IXR → BBI → BLR', 'DEL → LKO → BOM → COK',
       'CCU → NAG → BLR', 'BLR → NAG → DEL', 'CCU → BLR',
       'BLR → BOM → DEL', 'DEL → BOM → COK', 'DEL → BLR → COK',
       'MAA → CCU', 'CCU → BOM → BLR', 'DEL → AMD → BOM → COK',
       'DEL → PNQ → COK', 'DEL → CCU → BOM → COK', 'BLR → COK → DEL',
       'DEL → IDR → BOM → COK', 'DEL → LKO → COK',
       'CCU → GAU → DEL → BLR', 'DEL → NAG → BOM → COK',
       'CCU → MAA → BLR', 'DEL → HYD → COK', 'CCU → HYD → BLR',
       'DEL → COK', 'CCU → DEL → BLR', 'BLR → BOM → AMD → DEL',
       'BOM → DEL → HYD', 'DEL → MAA → COK', 'BOM → HYD',
       'DEL → BHO → BOM → COK', 'DEL → JAI → BOM → COK',
       'DEL → ATQ → BOM → COK', 'DEL → JDH → BOM → COK',
       'CCU → BBI → BOM → BLR', 'BLR → MAA → DEL',
       'DEL → GOI → BOM → COK', 'DEL → BDQ → BOM → COK',
       'CCU → JAI → BOM → BLR', 'CCU → BBI → BLR', 'BLR → HYD → DEL',
       'DEL → TRV → COK', 'CCU → IXR → DEL → BLR',
       'DEL → IXU → BOM → COK', 'CCU 

### 6. dep_time col:

In [32]:
df.dep_time.unique() # looks good

array(['22:20', '05:50', '09:25', '18:05', '16:50', '09:00', '18:55',
       '08:00', '08:55', '11:25', '09:45', '20:20', '11:40', '21:10',
       '17:15', '16:40', '08:45', '14:00', '20:15', '16:00', '14:10',
       '22:00', '04:00', '21:25', '21:50', '07:00', '07:05', '09:50',
       '14:35', '10:35', '15:05', '14:15', '06:45', '20:55', '11:10',
       '05:45', '19:00', '23:05', '11:00', '09:35', '21:15', '23:55',
       '19:45', '08:50', '15:40', '06:05', '15:00', '13:55', '05:55',
       '13:20', '05:05', '06:25', '17:30', '08:20', '19:55', '06:30',
       '14:05', '02:00', '09:40', '08:25', '20:25', '13:15', '02:15',
       '16:55', '20:45', '05:15', '19:50', '20:00', '06:10', '19:30',
       '04:45', '12:55', '18:15', '17:20', '15:25', '23:00', '12:00',
       '14:45', '11:50', '11:30', '14:40', '19:10', '06:00', '23:30',
       '07:35', '13:05', '12:30', '15:10', '12:50', '18:25', '16:30',
       '00:40', '06:50', '13:00', '19:15', '01:30', '17:00', '10:00',
       '19:35', '15:

* we can split dep_time for two columns one for hours and one for minutes then dropping it.
* Also we can creat dep_time as part of the day time like morning, afternoon, Evening, Night

In [33]:
# Creating dep_hour, and dep_minute columns from dep_time then dropping it.
df['dep_hour'] = df['dep_time'].str.split(':').str[0].astype('int')
df['dep_minute'] = df['dep_time'].str.split(':').str[1].astype('int')
df.drop('dep_time',axis=1,inplace=True)

In [34]:
# creating part of the day function
def par_of_the_day(hour):
    if 4 <= hour < 8:
        return 'Early Morning'
    elif 8 <= hour < 12:
        return 'Morning'
    elif 12 <= hour < 16:
        return 'Afternoon'
    elif 16 <= hour < 20:
        return 'Evening'
    elif 20 <= hour < 24:
        return 'Night'
    else:
        return 'Late Night'
    
# Creating dep_time as part of the day time
df['dep_time'] = df['dep_hour'].apply(par_of_the_day)

In [35]:
df['dep_time'].unique()

array(['Night', 'Early Morning', 'Morning', 'Evening', 'Afternoon',
       'Late Night'], dtype=object)

In [36]:
df.head()

airline    source destination                  route  arrival_time  \
0       IndiGo  Banglore       Delhi              BLR → DEL  01:10 22 Mar   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR         13:15   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK  04:25 10 Jun   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR         23:30   
4       IndiGo  Banglore       Delhi        BLR → NAG → DEL         21:35   

  duration total_stops additional_info  price  year  month  day  day_name  \
0   2h 50m    non-stop         No info   3897  2019      3   24    Sunday   
1   7h 25m     2 stops         No info   7662  2019      1    5  Saturday   
2      19h     2 stops         No info  13882  2019      9    6    Friday   
3   5h 25m      1 stop         No info   6218  2019     12    5  Thursday   
4   4h 45m      1 stop         No info  13302  2019      1    3  Thursday   

   holiday  season  dep_hour  dep_minute       dep_time  
0        1  Summer        22          20          Night  
1        1  Winter         5          50  Early Morning  
2        0   Rainy         9          25        Morning  
3        0  Winter        18           5        Evening  
4        0  Winter        16          50        Evening

### 7. arrival_time col:

In [37]:
# Set display options to show all rows
pd.set_option('display.max_rows', None)
df.arrival_time.value_counts()

19:00           412
21:00           360
19:15           333
16:10           154
12:35           122
20:45           112
22:30           111
18:50           110
22:50           104
11:20            95
19:50            94
21:20            88
12:20            87
22:25            77
21:05            72
23:35            71
20:05            69
14:05            68
08:35            66
12:00            63
19:45            58
23:00            57
07:15            56
12:55            55
13:15            54
11:25            54
18:10            54
10:40            53
15:30            53
19:40            53
08:45            52
22:00            52
10:10            51
22:35            50
22:55            50
19:35            50
12:10            49
16:20            49
18:15            49
21:50            49
07:45            49
09:30            49
13:55            49
11:50            49
10:05            48
08:50            48
10:35            48
23:15            46
22:20            45
13:20            44


* We split arrival_time to hour, minute, day, and month and then drop the arrival time.
* Also, we can check if the arrival day not before departure day to deal with it.
* further more we can make arrival time as part of the day time

In [38]:
# Creating arrival_hour, arrival_minute, arrival_day, and arrival_month columns.
df['arrival_day'] = df['arrival_time'].str.split(' ').str[1]
df['arrival_month'] = df['arrival_time'].str.split(' ').str[2]
df['arrival_time'] = df['arrival_time'].str.split(' ').str[0]
df['arrival_hour'] = df['arrival_time'].str.split(':').str[0].astype('int')
df['arrival_minute'] = df['arrival_time'].str.split(':').str[1].astype('int')
df.drop('arrival_time',axis=1,inplace=True)

In [39]:
df.head()

airline    source destination                  route duration  \
0       IndiGo  Banglore       Delhi              BLR → DEL   2h 50m   
1    Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   
2  Jet Airways     Delhi      Cochin  DEL → LKO → BOM → COK      19h   
3       IndiGo   Kolkata    Banglore        CCU → NAG → BLR   5h 25m   
4       IndiGo  Banglore       Delhi        BLR → NAG → DEL   4h 45m   

  total_stops additional_info  price  year  month  ...  day_name holiday  \
0    non-stop         No info   3897  2019      3  ...    Sunday       1   
1     2 stops         No info   7662  2019      1  ...  Saturday       1   
2     2 stops         No info  13882  2019      9  ...    Friday       0   
3      1 stop         No info   6218  2019     12  ...  Thursday       0   
4      1 stop         No info  13302  2019      1  ...  Thursday       0   

   season dep_hour  dep_minute       dep_time arrival_day arrival_month  \
0  Summer       22          20          Night          22           Mar   
1  Winter        5          50  Early Morning         NaN           NaN   
2   Rainy        9          25        Morning          10           Jun   
3  Winter       18           5        Evening         NaN           NaN   
4  Winter       16          50        Evening         NaN           NaN   

  arrival_hour  arrival_minute  
0            1              10  
1           13              15  
2            4              25  
3           23              30  
4           21              35  

[5 rows x 21 columns]

In [40]:
# checking the null values Percentage in arrival day and month columns
round(df[['arrival_month', 'arrival_day']].isnull().mean() * 100, 2)

arrival_month    60.41
arrival_day      60.41
dtype: float64

* As not all rows have arrival day and month we will take only ones have

In [41]:
# Creating data frame df_2 which contain arrival day and month 
df_2 = df[~df['arrival_month'].isnull()]

In [42]:
# checkin for arrival month unique values
df_2['arrival_month'].unique()

array(['Mar', 'Jun', 'May', 'Apr'], dtype=object)

In [43]:
# mapping arrival month with month numbers 
month = {'Mar':3,'Apr':4,'May':5,'Jun':6}
df_2['arrival_month'] = df_2['arrival_month'].map(month).astype('int')

In [44]:
df_2['arrival_month'].unique()

array([3, 6, 5, 4])

In [45]:
df_2.shape

(4142, 21)

In [46]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4142 entries, 0 to 10453
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   airline          4142 non-null   object
 1   source           4142 non-null   object
 2   destination      4142 non-null   object
 3   route            4142 non-null   object
 4   duration         4142 non-null   object
 5   total_stops      4142 non-null   object
 6   additional_info  4142 non-null   object
 7   price            4142 non-null   int64 
 8   year             4142 non-null   int64 
 9   month            4142 non-null   int64 
 10  day              4142 non-null   int64 
 11  day_name         4142 non-null   object
 12  holiday          4142 non-null   int32 
 13  season           4142 non-null   object
 14  dep_hour         4142 non-null   int32 
 15  dep_minute       4142 non-null   int32 
 16  dep_time         4142 non-null   object
 17  arrival_day      4142 non-null  

In [47]:
# conver arrival day into intger
df_2.arrival_day = df_2.arrival_day.astype('int')

In [48]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4142 entries, 0 to 10453
Data columns (total 21 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   airline          4142 non-null   object
 1   source           4142 non-null   object
 2   destination      4142 non-null   object
 3   route            4142 non-null   object
 4   duration         4142 non-null   object
 5   total_stops      4142 non-null   object
 6   additional_info  4142 non-null   object
 7   price            4142 non-null   int64 
 8   year             4142 non-null   int64 
 9   month            4142 non-null   int64 
 10  day              4142 non-null   int64 
 11  day_name         4142 non-null   object
 12  holiday          4142 non-null   int32 
 13  season           4142 non-null   object
 14  dep_hour         4142 non-null   int32 
 15  dep_minute       4142 non-null   int32 
 16  dep_time         4142 non-null   object
 17  arrival_day      4142 non-null  

In [49]:
# Indices of rows that have arrival day or month before departure day or month
df_2[(df_2['month'] > df_2['arrival_month']) | (df_2['day'] > df_2['arrival_day'])]

airline    source destination  \
0                    IndiGo  Banglore       Delhi   
2               Jet Airways     Delhi      Cochin   
6               Jet Airways  Banglore       Delhi   
7               Jet Airways  Banglore       Delhi   
8               Jet Airways  Banglore       Delhi   
13              Jet Airways   Kolkata    Banglore   
17              Jet Airways     Delhi      Cochin   
18                Air India     Delhi      Cochin   
21                Air India  Banglore       Delhi   
24              Jet Airways   Kolkata    Banglore   
25              Jet Airways     Delhi      Cochin   
26                   IndiGo     Delhi      Cochin   
51              Jet Airways   Kolkata    Banglore   
58        Multiple carriers     Delhi      Cochin   
59                  Vistara    Mumbai   Hyderabad   
71              Jet Airways  Banglore       Delhi   
80              Jet Airways   Kolkata    Banglore   
82              Jet Airways   Kolkata    Banglore   
85              Jet Airways  Banglore       Delhi   
92              Jet Airways  Banglore       Delhi   
99              Jet Airways     Delhi      Cochin   
101             Jet Airways   Kolkata    Banglore   
103               Air India   Kolkata    Banglore   
108               Air India     Delhi      Cochin   
110             Jet Airways  Banglore       Delhi   
112                 Vistara   Kolkata    Banglore   
113             Jet Airways     Delhi      Cochin   
116               Air India     Delhi      Cochin   
122             Jet Airways   Kolkata    Banglore   
123               Air India     Delhi      Cochin   
126             Jet Airways   Kolkata    Banglore   
127             Jet Airways   Kolkata    Banglore   
140       Multiple carriers     Delhi      Cochin   
146             Jet Airways     Delhi      Cochin   
155               Air India   Kolkata    Banglore   
158             Jet Airways   Kolkata    Banglore   
159             Jet Airways     Delhi      Cochin   
160             Jet Airways   Kolkata    Banglore   
162               Air India   Kolkata    Banglore   
174             Jet Airways   Kolkata    Banglore   
175             Jet Airways   Kolkata    Banglore   
178             Jet Airways   Kolkata    Banglore   
180             Jet Airways     Delhi      Cochin   
183             Jet Airways   Kolkata    Banglore   
194             Jet Airways  Banglore       Delhi   
196             Jet Airways   Kolkata    Banglore   
224             Jet Airways     Delhi      Cochin   
234                  IndiGo     Delhi      Cochin   
236             Jet Airways     Delhi      Cochin   
246             Jet Airways     Delhi      Cochin   
251               Air India     Delhi      Cochin   
260                  IndiGo  Banglore       Delhi   
267               Air India   Kolkata    Banglore   
269             Jet Airways  Banglore       Delhi   
270             Jet Airways   Kolkata    Banglore   
274               Air India     Delhi      Cochin   
278               Air India     Delhi      Cochin   
280             Jet Airways   Kolkata    Banglore   
284                  IndiGo   Kolkata    Banglore   
289               Air India  Banglore       Delhi   
311             Jet Airways     Delhi      Cochin   
318             Jet Airways     Delhi      Cochin   
321             Jet Airways     Delhi      Cochin   
323             Jet Airways  Banglore       Delhi   
336             Jet Airways     Delhi      Cochin   
342             Jet Airways   Kolkata    Banglore   
344               Air India     Delhi      Cochin   
347             Jet Airways     Delhi      Cochin   
348               Air India   Kolkata    Banglore   
353               Air India  Banglore       Delhi   
355                  IndiGo  Banglore       Delhi   
363               Air India  Banglore       Delhi   
377             Jet Airways   Kolkata    Banglore   
379             Jet Airways  Banglore       Delhi   
380       Multiple carriers     Delhi      Co

In [50]:
indices = df_2[(df_2['month'] > df_2['arrival_month']) | (df_2['day'] > df_2['arrival_day'])].index
indices

Int64Index([    0,     2,     6,     7,     8,    13,    17,    18,    21,
               24,
            ...
            10421, 10429, 10430, 10433, 10434, 10438, 10441, 10447, 10448,
            10453],
           dtype='int64', length=2037)

In [51]:
# drop indices form original data frame
df.drop(index= indices, axis=0, inplace=True)

In [52]:
# droping arrival day and arrival month columns as they contain 60% null values
df.drop(['arrival_day','arrival_month'],axis=1,inplace=True)

In [53]:
#checking for duplication in data 
df.duplicated().sum()

2

In [54]:
# Drop duplicates and reset the index
df = df.drop_duplicates().reset_index(drop=True)

In [55]:
df.duplicated().sum()

0

In [56]:
# using part of the day function to create new arrival time as part of the day
df['arrival_time'] = df['arrival_hour'].apply(par_of_the_day)

In [57]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8423 entries, 0 to 8422
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   airline          8423 non-null   object
 1   source           8423 non-null   object
 2   destination      8423 non-null   object
 3   route            8423 non-null   object
 4   duration         8423 non-null   object
 5   total_stops      8423 non-null   object
 6   additional_info  8423 non-null   object
 7   price            8423 non-null   int64 
 8   year             8423 non-null   int64 
 9   month            8423 non-null   int64 
 10  day              8423 non-null   int64 
 11  day_name         8423 non-null   object
 12  holiday          8423 non-null   int32 
 13  season           8423 non-null   object
 14  dep_hour         8423 non-null   int32 
 15  dep_minute       8423 non-null   int32 
 16  dep_time         8423 non-null   object
 17  arrival_hour     8423 non-null   

In [58]:
df.tail()

airline    source destination                  route duration  \
8418     Air Asia   Kolkata    Banglore              CCU → BLR   2h 30m   
8419    Air India   Kolkata    Banglore              CCU → BLR   2h 35m   
8420  Jet Airways  Banglore       Delhi              BLR → DEL       3h   
8421      Vistara  Banglore       Delhi              BLR → DEL   2h 40m   
8422    Air India     Delhi      Cochin  DEL → GOI → BOM → COK   8h 20m   

     total_stops additional_info  price  year  month  day   day_name  holiday  \
8418    non-stop         No info   4107  2019      9    4  Wednesday        0   
8419    non-stop         No info   4145  2019      4   27   Saturday        1   
8420    non-stop         No info   7229  2019      4   27   Saturday        1   
8421    non-stop         No info  12648  2019      1    3   Thursday        0   
8422     2 stops         No info  11753  2019      9    5   Thursday        0   

      season  dep_hour  dep_minute dep_time  arrival_hour  arrival_minute  \
8418   Rainy        19          55  Evening            22              25   
8419  Summer        20          45    Night            23              20   
8420  Summer         8          20  Morning            11              20   
8421  Winter        11          30  Morning            14              10   
8422   Rainy        10          55  Morning            19              15   

     arrival_time  
8418        Night  
8419        Night  
8420      Morning  
8421    Afternoon  
8422      Evening

### 8. duration col:

In [59]:
df.duration.value_counts()

2h 50m     503
1h 30m     374
2h 45m     327
2h 35m     318
2h 55m     305
3h         252
2h 30m     209
2h 20m     207
2h 40m     143
2h 15m     135
1h 25m     133
2h 25m      96
3h 15m      85
3h 5m       78
9h 30m      76
7h 15m      74
13h 30m     74
5h 30m      72
7h 5m       65
9h          64
7h 35m      63
7h 30m      63
9h 50m      63
8h          61
5h          61
1h 20m      61
11h 45m     58
8h 40m      58
11h         58
13h 20m     57
5h 15m      56
7h 20m      51
4h 55m      51
11h 30m     47
5h 45m      47
10h         46
6h 35m      46
10h 30m     45
5h 20m      44
10h 15m     44
6h 15m      44
8h 25m      43
12h         43
12h 30m     42
9h 15m      42
10h 25m     41
9h 20m      41
3h 10m      40
12h 15m     38
9h 35m      37
14h 25m     37
6h 25m      37
11h 15m     36
13h 15m     36
4h 50m      35
13h         35
4h 30m      35
12h 45m     34
15h 10m     33
12h 10m     33
8h 55m      32
16h 5m      32
12h 20m     31
8h 30m      31
11h 5m      31
5h 25m      31
14h       

* We can found that there is strange value for duration of 5m we are going to check it out
* Also I am going to change duration value for hour and minutes to integer minutes values.

In [60]:
df[df['duration'] == '5m']

airline  source destination                  route duration  \
5145  Air India  Mumbai   Hyderabad  BOM → GOI → PNQ → HYD       5m   

     total_stops additional_info  price  year  month  day day_name  holiday  \
5145     2 stops         No info  17327  2019      6    3   Monday        0   

     season  dep_hour  dep_minute dep_time  arrival_hour  arrival_minute  \
5145  Rainy        16          50  Evening            16              55   

     arrival_time  
5145      Evening

* From the above we can found that there are 2 stops for the flight from Mumbai to hyderabad on 3 june and the duration only 5 minute from 16:50 departure time to 16:55 arrival time this row is corrupted we are going to drop it. 

In [61]:
# selcting corrupted index to be dropped
corrupted_index = df[df['duration'] == '5m'].index

In [62]:
# dropping corrupted index
df.drop(index = corrupted_index, axis=0,inplace=True)

In [63]:
# reseting index
df.reset_index(drop=True, inplace=True)

In [64]:
df.head()

airline    source destination                  route duration  \
0          Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR   7h 25m   
1             IndiGo   Kolkata    Banglore        CCU → NAG → BLR   5h 25m   
2             IndiGo  Banglore       Delhi        BLR → NAG → DEL   4h 45m   
3           SpiceJet   Kolkata    Banglore              CCU → BLR   2h 25m   
4  Multiple carriers     Delhi      Cochin        DEL → BOM → COK   7h 50m   

  total_stops additional_info  price  year  month  day  day_name  holiday  \
0     2 stops         No info   7662  2019      1    5  Saturday        1   
1      1 stop         No info   6218  2019     12    5  Thursday        0   
2      1 stop         No info  13302  2019      1    3  Thursday        0   
3    non-stop         No info   3873  2019      6   24    Monday        0   
4      1 stop         No info   8625  2019      5   27    Monday        0   

   season  dep_hour  dep_minute       dep_time  arrival_hour  arrival_minute  \
0  Winter         5          50  Early Morning            13              15   
1  Winter        18           5        Evening            23              30   
2  Winter        16          50        Evening            21              35   
3   Rainy         9           0        Morning            11              25   
4  Summer        11          25        Morning            19              15   

  arrival_time  
0    Afternoon  
1        Night  
2        Night  
3      Morning  
4      Evening

In [65]:
# creating function for converting duration column to minutes integer values.
def convert_duration(x):
    if 'h' in x and 'm' in x:
        hours = int(x.split('h')[0])
        minutes = int(x.split(' ')[1].split('m')[0])
        return hours * 60 + minutes
    elif 'h' in x:
        return int(x.split('h')[0]) * 60
    elif 'm' in x:
        return int(x.split('m')[0])
    else:
        return 0

In [66]:
# appling convert_duration to duration column to get duration in minutes
df['duration'] = df['duration'].apply(convert_duration)

### 9. total_stops col:

In [67]:
df.total_stops.unique() # looks good

array(['2 stops', '1 stop', 'non-stop', '3 stops'], dtype=object)

In [68]:
# changing the the value of total_stops to numerical by mapping 
dict_total_stops = {'non-stop':0,'1 stop':1,'2 stops':2,'3 stops':3}
df.total_stops = df.total_stops.map(dict_total_stops)
df.total_stops.unique()

array([2, 1, 0, 3], dtype=int64)

* I would like to compare total_stops with the reoute columns

In [69]:
stops = []
route_list = df.route.str.split('→').to_list()
for i in route_list:
    number_of_stop = len(i) - 2
    stops.append(number_of_stop)

In [70]:
min(stops), max(stops)

(0, 3)

* when comparing stops obtaind from route column to total_stops:
   - minimum stops optain from route = 0 which is same as no-stop from total_stops.
   - maximum stops optain from route = 3 which is same as 3 stops from total_stops.
* from that we can asume that there is no contradictory between route & total_stops.

### 10. additional_info col:

In [71]:
df.additional_info.unique()

array(['No info', 'In-flight meal not included',
       'No check-in baggage included', '1 Short layover', 'No Info',
       '1 Long layover', 'Change airports', 'Business class'],
      dtype=object)

* There is typing issues like 'No info' and 'No Info' we are going to change all values to lower case.

In [72]:
# changing additional_info values to lower case
df['additional_info'] = df['additional_info'].str.lower()

In [73]:
df.additional_info.unique()

array(['no info', 'in-flight meal not included',
       'no check-in baggage included', '1 short layover',
       '1 long layover', 'change airports', 'business class'],
      dtype=object)

In [74]:
#  checking the value count for additional_info
df.additional_info.value_counts()/df.shape[0] *100

no info                         79.886013
in-flight meal not included     16.385657
no check-in baggage included     3.526478
1 long layover                   0.083116
change airports                  0.059368
business class                   0.047495
1 short layover                  0.011874
Name: additional_info, dtype: float64

* Almost 80% of the additional_info columns are no info or null values so it is useless column and better to be dropped.

In [75]:
# dropping additional_info useless column
df.drop('additional_info',axis=1,inplace=True)

In [76]:
# Checking for duplicate after dropping additional_info column
df.duplicated().sum()

0

### 11. price col:

In [77]:
df.price.value_counts() # looks good

10844    179
7229     161
10262    158
4804     147
4823     131
3943      92
3597      86
12898     85
3841      84
14781     79
3873      68
14714     66
5678      65
4174      63
8016      60
6442      60
3100      58
2754      56
14151     56
4544      51
4423      51
14571     50
4668      49
12373     46
3543      45
15129     42
13587     40
3419      40
8040      39
2227      39
4049      38
9646      38
6961      37
8586      36
4995      35
8372      35
16079     35
4878      33
9663      33
5228      32
2017      31
13377     31
1965      31
6093      30
4030      29
4148      29
5403      28
3625      28
3383      27
7832      26
7480      26
13727     26
6216      25
14871     25
14388     25
13941     25
9345      25
5613      25
13014     25
3898      25
13067     25
3850      25
12192     25
4409      24
6144      24
5769      24
10368     24
5192      23
9134      23
4226      23
5198      23
7408      23
8266      23
11134     22
5224      22
15554     22
14231     22

In [78]:
df.head()

airline    source destination                  route  duration  \
0          Air India   Kolkata    Banglore  CCU → IXR → BBI → BLR       445   
1             IndiGo   Kolkata    Banglore        CCU → NAG → BLR       325   
2             IndiGo  Banglore       Delhi        BLR → NAG → DEL       285   
3           SpiceJet   Kolkata    Banglore              CCU → BLR       145   
4  Multiple carriers     Delhi      Cochin        DEL → BOM → COK       470   

   total_stops  price  year  month  day  day_name  holiday  season  dep_hour  \
0            2   7662  2019      1    5  Saturday        1  Winter         5   
1            1   6218  2019     12    5  Thursday        0  Winter        18   
2            1  13302  2019      1    3  Thursday        0  Winter        16   
3            0   3873  2019      6   24    Monday        0   Rainy         9   
4            1   8625  2019      5   27    Monday        0  Summer        11   

   dep_minute       dep_time  arrival_hour  arrival_minute arrival_time  
0          50  Early Morning            13              15    Afternoon  
1           5        Evening            23              30        Night  
2          50        Evening            21              35        Night  
3           0        Morning            11              25      Morning  
4          25        Morning            19              15      Evening

In [79]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8422 entries, 0 to 8421
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   airline         8422 non-null   object
 1   source          8422 non-null   object
 2   destination     8422 non-null   object
 3   route           8422 non-null   object
 4   duration        8422 non-null   int64 
 5   total_stops     8422 non-null   int64 
 6   price           8422 non-null   int64 
 7   year            8422 non-null   int64 
 8   month           8422 non-null   int64 
 9   day             8422 non-null   int64 
 10  day_name        8422 non-null   object
 11  holiday         8422 non-null   int32 
 12  season          8422 non-null   object
 13  dep_hour        8422 non-null   int32 
 14  dep_minute      8422 non-null   int32 
 15  dep_time        8422 non-null   object
 16  arrival_hour    8422 non-null   int32 
 17  arrival_minute  8422 non-null   int32 
 18  arrival_

In [80]:
#checking for duplication in data 
df.duplicated().sum()

0

In [81]:
df.shape

(8422, 19)

In [82]:
# saving file after cleaning process.
df.to_csv('cleaned_data.csv',index=False)

## Data Cleaning & Feature Engineering Summery:

#### - The data set shape change from (10683, 11) to (8422 entries, 19 columns) after cleaning & Feature Engineering process.
#### - Maintaing features columns name to be all lower case

#### - The summery of the cleaning & Feature Engineering for each features as follows:
##### 1. **`Date_of_Journey`:**
    * Date_of_Journey data type changed from object to datetime.
    * date_of_journey as date time it is not useful but we create new features from it like year, month, day, day name, then dropping the 'Date_of_Journey column.
    * also we create new features holiday with 1 refering to holiday and 0 for not holiday. after checking from internet:
          * The standard working days in India is Monday to Friday then weekends days are Saturday and Sunday
          * In India , National holidays vary according to its local state but there’re ones which applied over the whole country like: 
            - January 26 -> Republic Day Celebrates the 1950 adoption of the Constitution of India
            - August 15  -> Independence Day Celebrates the 1947 Independence from the British rule
            - October 2  -> Gandhi Jayanti Honors Mahatma Gandhi, father of the nation, who was born on October 2, 1869
            - Reference: https://en.wikipedia.org/wiki/Public_holidays_in_India  
             
    * From the Month Features we Create New features Seasons. The climate of India consists of a wide range of weather conditions across a vast geographic scale and varied topography. but the main seasons are:
           - Winter, occurring from December to February.
           - Summer or pre-monsoon season, lasting from March to May.
           - Monsoon or rainy season, lasting from June to September. 
           - Post-monsoon or autumn season, lasting from October to November.
           - Reference: https://en.wikipedia.org/wiki/Climate_of_India#Seasons
 ##### 2. **`destination`:**
   * We noticed that New Delhi and Delhi are both referring to the same destination with the same route abbreviation DEL. I replace New Delhi with Delhi value.  
 ##### 3. **`dep_time`:** 
   * split dep_time for two new columns one for hours and one for minutes then dropping it.
   * Also Creat dep_time as part of the day time like morning, afternoon, Evening, Night from dep_hour    
 ##### 4. **`arrival_time`:** 
   * We split arrival_time to hour, minute, day, and month columns and then drop the arrival time.
   * Also, we check if the arrival day not before departure day and deal with it by dropping 2037 indices of corrupted departure and arrival month and date.
   * droping arrival day and arrival month columns as they contain over 60% null values.
   * further more Create arrival time as part of the day time like (Early morning, morning, afternoon, Evening, Night, Late Night) from arrival_hour. 
 ##### 5. **`duration`:**
   * We found that there is strange value for duration of 5m we drop it.
   * Also change duration value from hour and minutes to integer minutes values. 
 ##### 6. **`additional_info`:**
   * There is typing issues like 'No info' and 'No Info' we handle it by changing all values to lower case.
   * Almost 80% of the additional_info columns are no info or null values so we dropped it.
 ##### 7. **`total_stops`:**
   * when comparing stops obtaind from route column to total_stops:
       - minimum stops optain from route = 0 which is same as no-stop from total_stops.
       - maximum stops optain from route = 3 which is same as 3 stops from total_stops.
   * from that we can asume that there is no contradictory between route & total_stops.
   * Also, changing the values of total stops to numerical values.
     
#### -  saving file as cleaned_data.csv after cleaning & Features Engineering process for next project step.